src/sonic-frr/patch/0007-BFD-Enhancement.patch

```diff
@@ -0,0 +1,921 @@
+ 		return "Down";
+ 	case BFD_STATUS_UP:
+ 		return "Up";
++	case BFD_STATUS_ADMIN_DOWN:
++		return "Admin Down";
+ 	case BFD_STATUS_UNKNOWN:
+ 	default:
+ 		return "Unknown";
+diff --git a/lib/bfd.h b/lib/bfd.h
+index e4781f4ea..7f5d11150 100644
+--- a/lib/bfd.h
++++ b/lib/bfd.h
+@@ -51,9 +51,17 @@ struct bfd_gbl {
+ #define BFD_FLAG_BFD_CBIT_ON (1 << 3) /* Peer registered with CBIT set to on */
+ #define BFD_FLAG_BFD_CHECK_CONTROLPLANE (1 << 4) /* BFD and controlplane daemon are linked */
+ 
+-#define BFD_STATUS_UNKNOWN (1 << 0) /* BFD session status never received */
+-#define BFD_STATUS_DOWN    (1 << 1) /* BFD session status is down */
+-#define BFD_STATUS_UP      (1 << 2) /* BFD session status is up */
++#define BFD_STATUS_UNKNOWN    (1 << 0) /* BFD session status never received */
++#define BFD_STATUS_DOWN       (1 << 1) /* BFD session status is down */
++#define BFD_STATUS_UP         (1 << 2) /* BFD session status is up */
++#define BFD_STATUS_ADMIN_DOWN (1 << 3) /* BFD session is admin down */
++
++#define BFD_SET_CLIENT_STATUS(current_status, new_status)		  \
++	do {								  \
++		(current_status) =					  \
++			(((new_status) == BFD_STATUS_ADMIN_DOWN) ?	  \
++					  BFD_STATUS_DOWN : (new_status));\
++	} while (0)
+ 
+ enum bfd_sess_type {
+ 	BFD_TYPE_NOT_CONFIGURED,
+diff --git a/lib/command.h b/lib/command.h
+index 137d3748a..f6fe9af1b 100644
+--- a/lib/command.h
++++ b/lib/command.h
+@@ -398,6 +398,7 @@ struct cmd_node {
+ #define WATCHFRR_STR "watchfrr information\n"
+ #define ZEBRA_STR "Zebra information\n"
+ #define FILTER_LOG_STR "Filter Logs\n"
++#define BFD_STR "BFD information\n"
+ 
+ #define CMD_VNI_RANGE "(1-16777215)"
+ #define CONF_BACKUP_EXT ".sav"
+diff --git a/ospf6d/ospf6_bfd.c b/ospf6d/ospf6_bfd.c
+index f0500601b..4e7a0050a 100644
+--- a/ospf6d/ospf6_bfd.c
++++ b/ospf6d/ospf6_bfd.c
+@@ -236,7 +236,7 @@ static int ospf6_bfd_interface_dest_update(ZAPI_CALLBACK_ARGS)
+ 			continue;
+ 
+ 		old_status = bfd_info->status;
+-		bfd_info->status = status;
++		BFD_SET_CLIENT_STATUS(bfd_info->status, status);
+ 		monotime(&tv);
+ 		bfd_info->last_update = tv.tv_sec;
+ 
+diff --git a/ospfd/ospf_bfd.c b/ospfd/ospf_bfd.c
+index 05ec4991e..b9e78f4cd 100644
+--- a/ospfd/ospf_bfd.c
++++ b/ospfd/ospf_bfd.c
+@@ -262,7 +262,7 @@ static int ospf_bfd_interface_dest_update(ZAPI_CALLBACK_ARGS)
+ 			continue;
+ 
+ 		old_status = bfd_info->status;
+-		bfd_info->status = status;
++		BFD_SET_CLIENT_STATUS(bfd_info->status, status);
+ 		monotime(&tv);
+ 		bfd_info->last_update = tv.tv_sec;
+ 
+diff --git a/pimd/pim_bfd.c b/pimd/pim_bfd.c
+index 87d0f9fa2..01a798085 100644
+--- a/pimd/pim_bfd.c
++++ b/pimd/pim_bfd.c
+@@ -270,7 +270,7 @@ static int pim_bfd_interface_dest_update(ZAPI_CALLBACK_ARGS)
+ 			continue;
+ 		}
+ 		old_status = bfd_info->status;
+-		bfd_info->status = status;
++		BFD_SET_CLIENT_STATUS(bfd_info->status, status);
+ 		monotime(&tv);
+ 		bfd_info->last_update = tv.tv_sec;
+ 
```

Comment:
Why do we need this change?

Reply:
Scenarios where this code change is required:

BFD is un-configured from BGP at remote end.
Neighbour BFD sends ADMIN_DOWN state, but BFD on local side will send
DOWN to BGP, resulting in BGP session DOWN.
Removing BFD session administratively shouldn't bring DOWN BGP session
at local or remote.

BFD is un-configured from BGP or shutdown locally.
BFD will send state DOWN to BGP resulting in BGP session DOWN.
(This is akin to saying do not use BFD for BGP)
Removing BFD session administratively shouldn't bring DOWN BGP session at
local or remote.